Changlog for v0.23.0 (#1567)
simpeg/simpeg: docs/content/release/0.23.0-notes.rst

.. _0.23.0_notes:

===========================
SimPEG 0.23.0 Release Notes
===========================

November 1st, 2024

.. contents:: Highlights
    :depth: 3

Updates
=======

New features
------------

Full support for Numpy 2.0
^^^^^^^^^^^^^^^^^^^^^^^^^^

With this release SimPEG is fully compatible with Numpy 2.0.

Augmented receivers for airborne NSEM
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> We added new type of receivers intended to be used for airborne natural-source

Reviewer: ```suggestion
We added new types of receivers intended to be used for airborne natural-source
```

EM simulations:

- :class:`~simpeg.electromagnetics.natural_source.receivers.Impedance`
- :class:`~simpeg.electromagnetics.natural_source.receivers.Admittance`
- :class:`~simpeg.electromagnetics.natural_source.receivers.ApparentConductivity`
- :class:`~simpeg.electromagnetics.natural_source.receivers.Tipper`

They extend the type of airborne NSEM surveys that can be simulated in SimPEG.

See https://github.com/simpeg/simpeg/pull/1454 for more details.

Automatic selection of Solver
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> This release includes a new
> :func:`~simpeg.utils.solver_utils.get_default_solver` function that
> automatically selects the best solver available based on our system.
> For example, it'll select Pardiso as the solver if you are running an Intel
> CPU and have PARDISO installed in your system. Alternatively, it can choose
> Mumps if you are using Apple silicon. If no fast solver is available, it'll
> choose the SciPy's ``SparseLU`` solver.
> 
> Moreover, simulations will use this function to get the default solver if no
> solver is being provided by the user, making easier to run efficient
> finite volume simulations.
> 

Reviewer: ```suggestion
This release includes a new
:func:`~simpeg.utils.solver_utils.get_default_solver` function that
automatically selects a solver available based on you have installed on
your system.
For example, it'll select ``Pardiso`` as the solver if you are running an Intel
CPU and have `pydiso` installed in your system. Alternatively, it can choose
``Mumps`` if you are using Apple silicon and the `python-mumps` package is installed.
If no fast solver is available, it'll choose the SciPy's ``SparseLU`` solver.

.. note::
    For those installing through `conda-forge`, ``conda install simpeg`` will now also grab
    better solvers for your system as well, consistent with `simpeg`'s solver priority.

Moreover, simulations will also use this function to get the default solver if no
solver is being provided by the user, making it easier to run efficient
finite volume simulations out of the box.
```

See https://github.com/simpeg/simpeg/pull/1511 for more information.

Support for magnetic gradiometry in Choclo-based magnetic simulations
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Now the magnetic simulation with ``engine="choclo"`` support forward modelling

Reviewer: ```suggestion
Now the magnetic simulations with ``engine="choclo"`` support forward modelling
```

> the magnetic gradiometry components and TMI derivatives. This fully extend the

Reviewer: ```suggestion
the magnetic gradiometry components and TMI derivatives. This fully extends the
```

See https://github.com/simpeg/simpeg/pull/1543 and
https://github.com/simpeg/simpeg/pull/1553 for more information.

Numba-based implementation of gravity and equivalent sources
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This release includes new implementations of the gravity and magnetic
equivalent sources using `Choclo <https://fatiando.org/choclo>`__'s kernels and
forward modelling functions and `Numba <https://numba.pydata.org/>`__ to
just-in-time compilations and parallelization, making them faster and more
memory efficient.

See https://github.com/simpeg/simpeg/pull/1552 and
https://github.com/simpeg/simpeg/pull/1527.

New ``UpdateIRLS`` directive
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

We have renamed the :class:`simpeg.directives.UpdateIRLS` directive for
applying the Iterative Re-weighted Least-Squares during sparse norm inversions,
that includes better argument names and an improved implementation.
See https://github.com/simpeg/simpeg/pull/1349 for more details.

Standardize arguments for active cells and random seeds
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

We have standardize the name of the arguments for active cells in meshes through

> all functions and methods in SimPEG. We deprecated the old

Reviewer: ```suggestion
all functions and methods in SimPEG. We have deprecated the old
```

names like ``indAct``, ``ind_active`` and ``indActive`` in favor of
``active_cells``.

> The arguments for passing random seeds has also been standardize to

Reviewer: ```suggestion
The arguments for passing random seeds have also been standardized to
```

``random_seed``. These arguments generalize a way to specify random states,
allowing to take seeds as integers or instances of
:class:`numpy.random.Generator`.

Upgraded dependencies
---------------------

In SimPEG v0.23.0 we dropped support for Python versions <= 3.9. Python 3.8 met
its end-of-life this year (October 2024). Python 3.10 is the minimum required
version for Numpy 2.1.0. To keep up with the latest updates in the scientific
Python ecosystem, we decided to set Python 3.10 as the minimum required version
for SimPEG as well.

> Moreover, we have increased the minimum required versions ``discretize``,

Reviewer: ```suggestion
Moreover, we have increased the minimum required versions of ``discretize``,
```
